torch/_inductor/kernel/flex_attention.py

```diff
@@ -1113,78 +1275,149 @@ def bwd_dkdv_inner(
-    for start_m in range(0, hi):
-        # Load LSE before computing qk to reduce pipeline stall.
-
-        qT = tl.load(qT_ptrs)
-        lse = tl.load(LSE + offs_m1)
-        qkT = tl.dot(k, qT)
-        if not PRESCALE_QK:
-            qkT *= SM_SCALE
-        # ~~~~~~~~~~~~~~~~~~~ Apply score modification  ~~~~~~~~~~~~~~~~~~~
+
+    if not IS_DIVISIBLE:
+        if hi >= 1:
+            for start_m in range(0, hi - 1):
+                dk, dv = bwd_dkdv_compute_block_mn(
+                    dk, dv, qT_ptrs, k, v, do_ptrs, DELTA, LSE, Q_LEN, KV_LEN,
+                    off_z, off_h, offs_n1, offs_m1,
+                    stride_qm, stride_qd, stride_dom, stride_dod,
+                    q_indices, sparse_q_num_blocks,
+                    MATMUL_PRECISION, RCP_LN2,
+                    {{gen_argdefs()}}, IS_FULL_BLOCKS
+                )
+                # Increment pointers.
+                offset = get_offset_for_next_block(
+                    start_m, q_indices, sparse_q_num_blocks,
+                    SPARSE_Q_BLOCK_SIZE, SPARSE_Q_MULTIPLE, BLOCK_M1
+                )
+
+                qT_ptrs += offset * stride_qm
+                do_ptrs += offset * stride_dom
+
+                offs_m1 += offset
+
+            dk, dv = bwd_dkdv_compute_block_mn(
+                dk, dv, qT_ptrs, k, v, do_ptrs, DELTA, LSE, Q_LEN, KV_LEN,
+                off_z, off_h, offs_n1, offs_m1,
+                stride_qm, stride_qd, stride_dom, stride_dod,
+                q_indices, sparse_q_num_blocks,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}}, IS_FULL_BLOCKS, True,
+            )
+    else:
+        for start_m in range(0, hi):
+            dk, dv = bwd_dkdv_compute_block_mn(
+                dk, dv, qT_ptrs, k, v, do_ptrs, DELTA, LSE, Q_LEN, KV_LEN,
+                off_z, off_h, offs_n1, offs_m1,
+                stride_qm, stride_qd, stride_dom, stride_dod,
+                q_indices, sparse_q_num_blocks,
+                MATMUL_PRECISION, RCP_LN2,
+                {{gen_argdefs()}}, IS_FULL_BLOCKS
+            )
+            # Increment pointers.
+            offset = get_offset_for_next_block(
+                start_m, q_indices, sparse_q_num_blocks,
+                SPARSE_Q_BLOCK_SIZE, SPARSE_Q_MULTIPLE, BLOCK_M1
+            )
+
+            qT_ptrs += offset * stride_qm
+            do_ptrs += offset * stride_dom
+
+            offs_m1 += offset
+
+    return dk, dv
+
+
+@triton.jit
+def bwd_dkdv_compute_block_mn(
+    dk, dv, qT_ptrs, k, v, do_ptrs, DELTA, LSE, Q_LEN, KV_LEN,
+    off_z, off_h, offs_n1, offs_m1,
+    stride_qm, stride_qd, stride_dom, stride_dod,
+    q_indices, sparse_q_num_blocks,
+    MATMUL_PRECISION, RCP_LN2,
+    {{gen_argdefs()}}, IS_FULL_BLOCKS, is_last_block=False,
+):
+    {{gen_defines() | indent_except_first(1) }}
+
+    # Load LSE before computing qk to reduce pipeline stall.
+    qT = tl.load(qT_ptrs, mask=offs_m1[None, :] < Q_LEN)
```

Comment:
Does applying these masks always affect perf? When I tested it it seemed to have a couple percentage points of perf impact.

Reply:
Yes, I run benchmark after removing all ```mask``` or ```boundary_check``` for ```tl.load``` and got the following numbers:
```
fwd speedup: 0.83
bwd speedup: 0.81
```
It shows bwd perf on par with divisible seqlen, and fwd perf still can't catch it. So it seems the if condition and mask applying in the ```fwd_compute_block_mn``` hurts performance a lot, but ```bwd_dq_compute_block_mn``` and ```bwd_dkdv_compute_block_mn``` doesn't hurt too much.

Reply:
Oh, I think we should guard and only do ```mask``` or ```boundary_check``` for non-divisible case, this can improve the divisible case's perf.